robot/board.py

```diff
@@ -50,23 +55,34 @@ def _send(self, message, retry=False):
                     raise ConnectionError(self._get_lc_error())
                 self._send(message, retry=True)  # Retry Recursively
 
-    def _recv(self, retry=False):
-
+    def _recv(self, _retry=False):
         """
         Receive a message from the robotd socket
-        :return: message
+        :return: the message
         """
-        # TODO split receieves by \n characters and return them one at a time.
-        # Currently this is mitigated by having a large receive buffer, but I'd rather
-        # we did it properly at some point
-        try:
-            return self.sock.recv(Board.RECV_BUFFER)
-        except (socket.timeout, BrokenPipeError, OSError):
-            if retry:
-                raise ConnectionError(self._get_lc_error())
-            else:
-                self._connect(self.sock_path)  # Reconnect
-                return self._recv(retry=True)  # Retry recursively
+
+        # TODO catch empty string receives as bad things
+        unended_text = b''
+
+        while not self._recv_buffer:
```

Comment:
This code block is very dense and seems rather overly complex.

I also think there's a bug in this logic. If we receive data which looks like: `b'foo\nbar'` then at line 82 we end up with `strings` being `[b'foo', b'bar']` and `unended_text` being `b'bar'`. Since that's a truthy value, we won't modify `_recv_buffer` and will loop around.

Whatever we next receive will correctly be appended to `unended_text`, though we'll have thrown away the contents of `strings`.

I think the logic we want looks something like this:
``` python
def receive():
    data = b''
    lines = []
    while True:
        data += receive_raw_from_socket_with_single_retry()

        if data:
            lines += data.split(b'\n')

            # An empty last line indicates the clean end of the data (why?)
            if lines[-1] == b'':
                return lines
            else:
                # Extract the incomplete last line as the prefix for the next
                # chunk of data we get
                data = lines.pop()
```

I think we can also tidy up the socket handling logic by adding a helper which looks something like:
``` python
def _socket_with_single_retry(self, handler)
    retryable_errors = (socket.timeout, BrokenPipeError, OSError)

    try:
        return handler(self.sock)
    except retryable_errors:
        pass

    # Retry once, need to reconnect first
    try:
        self._connect(self.sock_path)
    except FileNotFoundError:
        raise ConnectionError(self._get_lc_error())

    try:
        return handler(self.sock)
    except retryable_errors:
        raise ConnectionError(self._get_lc_error())
```

This can be used for all socket actions and then means that `receive_raw_from_socket_with_single_retry` is pretty simple (and could even be inlined):
``` python
def receive_raw_from_socket_with_single_retry():
    return self._socket_with_single_retry(
        lambda s: s.recv(Board.RECV_BUFFER),
    )
```

Thoughts?